collections/collections.go

```diff
@@ -0,0 +1,81 @@
+package collections
+
+import (
+	"errors"
+	"math"
+
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+)
+
+// ErrNotFound is returned when the provided key is not present in the StorageProvider.
+var ErrNotFound = errors.New("collections: not found")
+
+// ErrEncoding is returned when something fails during key or value encoding/decoding.
+var ErrEncoding = errors.New("collections: encoding error")
+
+// StorageProvider represents sdk.Context
+// it is used to avoid to reduce dependencies.
+type StorageProvider interface {
+	// KVStore returns a KVStore given its StoreKey.
+	KVStore(key storetypes.StoreKey) storetypes.KVStore
+}
+
+// Prefix defines a segregation namespace
+// for specific collections objects.
+type Prefix struct {
+	raw []byte // TODO(testinginprod): maybe add a humanized prefix?
+}
+
+// Bytes returns the raw Prefix bytes.
+func (n Prefix) Bytes() []byte { return n.raw }
+
+// NewPrefix returns a Prefix given the provided namespace identifier.
+// Prefixes of the same module must not start
+func NewPrefix[T interface{ int | string | []byte }](identifier T) Prefix {
+	i := any(identifier)
+	var prefix []byte
+	switch c := i.(type) {
+	case int:
+		if c > math.MaxUint8 || c < 0 {
+			panic("invalid integer prefix value: must be between 0 and 255")
+		}
+		prefix = []byte{uint8(c)}
+	case string:
+		prefix = []byte(c)
+	case []byte:
+		prefix = c // maybe copy?
+	}
+	return Prefix{raw: prefix}
+}
+
+// KeyEncoder defines a generic interface which is implemented
+// by types that are capable of encoding and decoding collections keys.
+type KeyEncoder[T any] interface {
+	// PutKey writes the key bytes into the buffer. Returns the number of
+	// bytes written.
+	// The bytes written must match the value returned by the Size method.
+	PutKey(buffer []byte, key T) (int, error)
+	// ReadKey reads from the provided bytes buffer to decode
+	// the key T. Returns the number of bytes read, the type T
+	// or an error in case of decoding failure.
+	ReadKey(buffer []byte) (int, T, error)
+	// Size returns the size of the key T in binary format.
+	Size(key T) int
+	// Stringify returns a string representation of T.
+	Stringify(key T) string
+	// KeyType returns an identifier for the key.
+	KeyType() string
+}
+
+// ValueEncoder defines a generic interface which is implemented
+// by types that are capable of encoding and decoding collection values.
+type ValueEncoder[T any] interface {
+	// Encode encodes the value T into bytes.
+	Encode(value T) ([]byte, error)
+	// Decode returns the type T given its bytes representation.
+	Decode(b []byte) (T, error)
+	// Stringify returns a string representation of T.
+	Stringify(value T) string
+	// ValueType returns the identifier for the type.
+	ValueType() string
+}
```

Comment:
Again, just to reiterate — `ValueEncoder[T any]` models the concept of an encoder that can process multiple encodable types, e.g. JSONEncoder, XMLEncoder, etc. It factually does not model specific encodable types, e.g Uint64Value, LengthPrefixedStringValue, etc. If you want to model specific encodable types, then ValueEncoder should not be generic in this way, it should be simple interface, basically like encoding.Binary{Marshaler, Unmarshaler}, which concrete types would implement.

Reply:
I'm not sure if you have had time to read the **API must be nice section**[ here](https://github.com/cosmos/cosmos-sdk/pull/14134#issuecomment-1339294031), but even if `concrete types would implement` should be the most idiomatic path, it's not good for API consumers as they're forced into type conversions because `Map` cannot be generic over `K,V` with `any`.

It's a trade-off, and I'm happy to disagree with you here, I think readability and usability are more important in this context.

Reply:
> Map cannot be generic over K, V with any

> For this API to be nice, and with a low migration cost, we should be able to put every key/value in the Map[K, V]. Meaning that Map[string, uint64] should be a viable collection type.

You say 

> The consequence of this is the fact that, since we cannot implement anything on top of golang primitive types and types from packages we don't own, we allow K and V to be unconstrained in the Map and shift the duty to Key/ValueEncoder to convert K and V into bytes.

Can you define a Map with a K that does not have a KeyEncoder which can convert a value of that type K to []byte? No. Map[K, ...] requires a KeyEncoder[K] for every key type K you want to use. Like you can't have a Map with K=time.Time unless you also have a KeyEncoder for time.Time. So K is not "unconstrained", every Map[K] requires there to be a corresponding KeyEncoder[K]. This isn't a discussion of UX or usability or readability of APIs or anything like that, it's a fundamental property of type systems. Is this not clear? Am I misunderstanding the intent here?

Reply:
> Map[string, uint64] should be a viable collection type.

Map[string, uint64] is a viable collection type if, and only if, there is some code which implements the methods of the Map interface for K=string and V=uint64. Defining Map with K and V as `any` doesn't do that, it doesn't solve the problem. Is this not clear?

Reply:
@peterbourgon I think there might have been a misunderstanding to what I was referring to, to reiterate: 
What you want is, `K` constrained, eg: `Map[K Key, V Value]` where `type Key interface { Encode(); Decode() }`.

This has a fundamental API UX issue for which I cannot use, let's say, `time.Time` but I need to do `type MyKeyImplementerTime time.Time` and implement `Key` methods on top of it.

My design takes a different perspective and says `K` can be anything as long as I provide a type `KeyEncoder[K]` who is delegated the duty to ser/de `K` to bytes. So the constraint is in reality transitive on `KeyEncoder[K]`, it's just a delegation of duty from `K` to `KeyEncoder[T]`.

Reply:
> Map[string, uint64] is a viable collection type if, and only if, there is some code which implements the methods of the Map interface for K=string and V=uint64

But Map is not an interface, it's a concrete type in collections parametrized over K and V.

Reply:
> > Map[string, uint64] is a viable collection type if, and only if, there is some code which implements the methods of the Map interface for K=string and V=uint64
> 
> ?? But Map is not an interface, it's a concrete type in collections parametrized over K and V.

Do you intend to write an e.g. Encode function that takes an arbitrary key of type any and returns a []byte? If yes, how can you write code which encodes any possible type to bytes in a way which is decodable? (Spoiler: you can't, not possible.) If no, then you need code specific to each concrete key type you intend to support, right?

Reply:
> > So the constraint is in reality transitive on KeyEncoder[K], it's just a delegation of duty from K to KeyEncoder[K].
> 
> How do you intend to support keys of type int, int8, int16, int32, int64, uint, uint8, uint16, uint64, float32, float64, string, byte, etc. etc.? Do you intend to define a KeyEncoder[x] where x = each of these concrete types, and write methods with discrete behaviors for each of those types?

Yes this is and always has been the intention. Not for all types, only for the most common ones. Then external consumers are free to implement their own `KeyEncoder` for their type, if unhappy with current implementation (backwards compat) or because it is not supported

Reply:
@peterbourgon I am concerned reading your comments that you simply have not taken the time to understand the contents of this PR, nor the voluminous responses others have given to you on this point. This type of engagement feels very unproductive for the developers contributing to this project. I can see that you are very smart, but it seems like people keep trying to explain the same thing to you and you’re not getting it. I encourage you to please spend more time trying to understand others, their code, and their points before responding. 

Regarding the contents of this thread, I feel like there were already plenty of responses to your questions before this specific sub-thread opened up, not to mention the rationale for this seeming pretty obvious from the contents of the PR (at least to me and I imagine others who have already worked with this code). So I ask you to please spend more time trying to understand before responding and please don't assume you're right. Having @testinginprod continue to respond to this feels unproductive. As far as I can tell the answers are already there and you just need to spend the time trying to grok. If you can take this to heart and take the effort to consider the different points of view people are trying to express, I think we can have much more productive discussions.